46.md

```diff
@@ -1,98 +1,174 @@
+
+### Example flow for signing an event
+
+Coming soon...
+
+### Example Oauth-like flow to create a new user account with Nsecbunker
+
+Coming soon...
 
-## Event payloads
+## Request Events `Multiple kinds`
 
-Event payloads are [NIP-04](04.md)-encrypted JSON blobs that look like JSONRPC messages (their format is specified inside the `.content` of the event formats below).
+Request events take the following form. In order simplify requests and payloads as much as possible, there are different event kinds for each command which help remote signers to understand the request and process the content properly.
 
-Events sent by the client to the remote signer have the following format:
+Clients can subscribe to events of `kind: 24136`, `kind: 24137`, and `kind: 24138` that have an "e" tag containing the ID of their request event to listen for the response, auth_challenge, or error to their command.
 
-```js
+```json
 {
-  "pubkey": "<client-key-hex>"
-  "kind": 24133,
-  "tags": [
-    ["p", "<signer-key-hex>"]
-  ],
-  "content": "nip04_encrypted_json({id: <random-string>, method: <see-below>, params: [array_of_strings]})",
-  ...
+    "id": <id>,
+    "kind": <kind for command>, // Each command has it's own kind
+    "pubkey": <local pubkey of the user>,
+    "content": <nip44(<request>)>,
+    "tags": [["p", <remote pubkey>]],
+    "created_at": <unix timestamp in seconds>,
 }
 ```
 
-And the events the remote signer sends to the client have the following format:
+| Kind    | Command                  | Content field (always NIP-44 encrypted)                                         | Data Type               |
+| ------- | ------------------------ | ------------------------------------------------------------------------------- | ----------------------- |
+| `24140` | `create_account`         | {username: <username>, domain: <domain>, <any other arbitrary key/value pairs>} | JSON stringified object |
+| `24141` | `connect`                | -                                                                               | null                    |
+| `24142` | `sign_event`             | nostr event to sign                                                             | JSON stringified object |
+| `24143` | `ping`                   | -                                                                               | null                    |
+| `24144` | `get_relays`             | -                                                                               | null                    |
+| `24145` | `nip04_encrypt`          | {third_party_pubkey: <third_party_pubkey>, content: <plaintext to encrypt>}     | JSON stringified object |
+| `24146` | `nip04_decrypt`          | {third_party_pubkey: <third_party_pubkey>, content: <ciphertext to decrypt>}    | JSON stringified object |
+| `24147` | `nip44_conversation_key` | third_party_pubkey                                                              | string                  |
```

Comment:
```suggestion
```

there was a lot of conversation around how to structure the nip44 calls for nip07 and we ultimately landed on not exposing the conversation key (incase we want to use that for other schemes in the future) and instead make the nip44 calls take and return arrays (to mitigate latency for multiple calls, especially since conversation key is no longer exposed). I think nip46 should also follow that pattern.

heres the pr: https://github.com/nostr-protocol/nips/pull/940

Reply:
Yeah – @staab flagged that discussion with me earlier. This might be reopening a painful wound but I'm not convinced that where the Nip44 calls in Nip07 stuff landed applies in the same way here. 

Remote signers are signing, encrypting, and decrypting for potentially many, many users instead of just one local user (as is the case in NIP-07). This means that you're shifting a potentially huge compute workload onto remote signers with the encryption stuff (since signing has to obviously stay remote). I completely understand why you all ended up doing away with the conversation key and allowing arrays for encryption the NIP-07 stuff, but I think it's probably more appropriate here to do the opposite. 

Asking a remote signer to decrypt your entire inbox each time your load up your DMs (which is how naive clients will do it) across thousands or eventually millions of users would require huge compute for remote signers, not to mention added network latency.

The idea of a rouge client doing something nefarious with your conversation key is a problem that, while I'm sure it'll happen eventually and is very bad, it's also something that is more easy to manage/solve with good UI/UX (remote signers can display warnings or limit the scope of usage of this method).

My main fear is that we'll end up with very few (read: centralized) remote signers if they have to be beasts of compute to manage all the remote work they have to do for clients.

Reply:
neutering these calls to only take single arguments doesn't really fix this concern though. clients will still request for the same number of decrypts/encrypts. if a signer wants to rate limit how many decryptions are done per minute, it can easily do that even if the requests are submitted by array by only returning the first few elements decrypted.

I don't think its fair to make these calls less functional just to protect one particular use case of a signer. resource allocation is something an nsecbunker needs to handle, and we shouldn't use specifications as a proxy to handle this. besides, this is very unfair to small personal signers which would greatly benefit from these calls to take arrays.

as far as centralization, the most decentralized remote signer model is if everyone has their own app that manages their signing rather than relying on any external service. by tailoring these calls for nsecbunker I feel like you are actually favoring more centralized solutions.

Reply:
also, if we expose conversation key (which is needed for efficiency if we only allow single encryption/decryption) then there is no possibility of using the conversation key as some sort of shared secret key between too users in the future. imo this could have really creative applications.